Use first non-empty Nav post as primary fallback for Nav block

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
In https://github.com/WordPress/gutenberg/pull/36724 we added the Page List block as a fallback for when the Nav block has no menu selected.

This PR iterates on that idea by preferring to use the first non-empty Navigation (i.e. `wp_navigation` Post) as the primary fallback. Only if a suitable Navigation Post is not found do we fallback (again!) to the Page List.

This aims to mirror how the fallback mechanic for `wp_nav_menu` works currently in Core (as [described in the Issue](https://github.com/WordPress/gutenberg/issues/36721)).

Closes https://github.com/WordPress/gutenberg/issues/36721

## How has this been tested?

Testing is as follows:

### Preparation

* Before testing use Fakerpress Plugin (or similar) to add a load of pages to your site. Some should be nested parent/child.
* Also use the Nav block to create one or two Menus on your site being sure to add some items. Make sure they are saved as Navigation post types (check by going to `Appearance -> Navigation`).

### Testing
- Open Site Editor.
- Remove any existing Nav block.
-  Add Nav block from scratch. Don't add any items!
- Save your empty Nav block and publish the post.
- Go to front of site.
- See that the first of your Navigation Menus is rendered by _default_ even though no Menu has been selected.
-  Remove _all of the items_ from the Navigation Post that is currently being used as the fallback - this should make it ineligible to be used as a fallback.
-  Go to front of site. Check that a different Navigation Post is being used as a fallback.
- Remove _all_ your Navigation Posts - that's right, ditch them all!
- Go to front of site. Check that it falls back to the Page List.


## Screenshots <!-- if applicable -->

## Types of changes
New feature (non-breaking change which adds functionality) 
## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Use non-empty Nav post as fallback

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -221,12 +221,38 @@ function( $block ) {
+
+		// Use non-empty Navigation if available.
+		if ( $navigation_post ) {
+			$fallback_blocks = parse_blocks( $navigation_post->post_content );
```
What happens here if we don't have any blocks? 🤔

talldan replied:
> What happens here if we don't have any blocks? 🤔

Isn't that handled by the `! empty` in the loop above? Alternatively you could `parse_blocks` in the loop and check how many there are before selecting the post as a candidate.

Also, just be aware of the code on line 205-212 that filters the parsed blocks, that will probably have to be applied here too.

getdave replied:
> Alternatively you could parse_blocks in the loop and check how many there are before selecting the post as a candidate.

I was wondering whether parsing would be an expensive operation.

getdave replied:
Ok I avoided parsing but added a `null` check on `blockName`.

talldan replied:
I don't think it'd be too expensive. 

If the navigation post is empty, the parser won't have anything to do and should skip it. If the navigation post has content, the parser will have something to do, but that will very likely be the chosen post, so the parsing was necessary.

But having said that, it should generally be ok as it is. I can't think of a situation where the content might have only line breaks.

talldan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -221,12 +221,38 @@ function( $block ) {
+		if ( $navigation_post ) {
+			$fallback_blocks = parse_blocks( $navigation_post->post_content );
+		} else {
+			$attributes['__unstableMaxPages'] = 4; // set value to be passed as context to Page List block.
```
What's the reason for using context? Feels like it should be handled as an attribute of Page List block rather than the navigation block.

It could be passed through in the `attrs` array a few lines below.

This also duplicates line 222, which is identical.

getdave replied:
You weren't to know but this has come up before. I was telling @adamziel that Page List doesn't appear to have attributes and that seems to be a deliberate architectural choice which I'm loathe to change.

See https://github.com/WordPress/gutenberg/pull/36724#discussion_r754330576.

talldan replied:
I don't think it is an architectural decision. It's just a new block that hasn't had many features developed for it yet, but there's plenty of scope to do that. See the enhancement requests:
https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BBlock%5D+Page+List%22+label%3A%22%5BType%5D+Enhancement%22

There are also some PRs open for the block that add attributes like this one - https://github.com/WordPress/gutenberg/pull/31471

getdave replied:
Ok in that case I will

- not change anything in _this_ PR (as the changes we're referring to we're introduced in https://github.com/WordPress/gutenberg/pull/36724).
- raise a follow up to make the Page List block use an attribute for `maxPages` and revert `maxPages` on the Nav block.

Sound good? cc @adamziel

talldan replied:
In this PR, we still need to address that this line of code is a duplicate to that on line 222.

getdave replied:
> we still need to address that this line of code is a duplicate to that on line 222.

✅ Done

talldan commented on packages/block-library/src/navigation/index.php:
```diff
@@ -221,12 +221,38 @@ function( $block ) {
 		$attributes['__unstableMaxPages'] = 4; // set value to be passed as context to Page List block.
 
-		$page_list_block = array(
-			'blockName' => 'core/page-list',
-			'attrs'     => array(),
+		$navigation_posts = get_posts(
+			array(
+				'post_type' => 'wp_navigation',
+			)
 		);
 
-		$inner_blocks = new WP_Block_List( array( $page_list_block ), $attributes );
+		$navigation_post = null;
+
+		// Pick first non-empty Navigation.
+		foreach ( $navigation_posts as $navigation_maybe ) {
+			if ( ! empty( $navigation_maybe->post_content ) ) {
+				$navigation_post = $navigation_maybe;
+				break;
+			}
+		}
+
+		// Use non-empty Navigation if available.
+		if ( $navigation_post ) {
+			$fallback_blocks = parse_blocks( $navigation_post->post_content );
+		} else {
+			$attributes['__unstableMaxPages'] = 4; // set value to be passed as context to Page List block.
+
+			// Requires wrapping array.
+			$fallback_blocks = array(
+				array(
+					'blockName' => 'core/page-list',
+					'attrs'     => array(),
+				),
+			);
+		}
+
+		$inner_blocks = new WP_Block_List( $fallback_blocks, $attributes );
```
This is subjective, but I feel like the code inside this if statement is now considerable enough to be extracted into a separate function.

getdave replied:
Extraction complete ✅

talldan reviewed (commented):
Looks good so far.

getdave committed:
Remove code duplication

Addresses https://github.com/WordPress/gutenberg/pull/36740#discussion_r754867803

getdave committed:
Extract finding non empty nav block to function

getdave committed:
Extract process of getting fallback to dedicated function

getdave committed:
Apply block prefix to functions

getdave committed:
Add additional safety check around parsing Nav blocks

getdave committed:
Extract function for removing null blocks and apply

getdave committed:
Check for empty parsed Nav blocks

draganescu commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,84 @@ function block_core_navigation_render_submenu_icon() {
+			'post_type' => 'wp_navigation',
+		)
+	);
+
```
What is the default ordering here? I mean, the first from what list? ID?

getdave replied:
Good point. It uses `[WP_Query](https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters)` so by default it's:

 - order `DESC`
 - orderby `date (post_date)`

We should probably look to emulate the same type of ordering as [`wp_get_nav_menus()`](https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/nav-menu.php#L613-L632) which defers to `get_terms()`.

Can't narrow down the order/order-by args as yet.

getdave replied:
Ok in [`wp_get_nav_menus` `orderby` is set to `name`](https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/nav-menu.php#L613-L617). Why that is useful I don't know...

And [`order` is set to `ASC` in `get_terms`](https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/taxonomy.php#L1234).

getdave replied:
https://github.com/WordPress/gutenberg/pull/36740/commits/2831ff35cf6373a75fac2a983e0e7b1a8ba35f23

draganescu commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,84 @@ function block_core_navigation_render_submenu_icon() {
+ * @return array the normalized parsed blocks.
+ */
+function block_core_navigation_normalize_parsed_blocks( $parsed_blocks ) {
+	$filtered = array_filter(
```
I understand the comment on what the function does, but when is this happening? Is it a bug? Is it a workaround?

getdave replied:
No this is just how the parser is supposed to work. I will add a comment to clarify.

getdave replied:
Added https://github.com/WordPress/gutenberg/pull/36740/commits/0ca7eb188660bb46253e2dd9844e94ca189235f4

carolinan commented:
I did a manual test by following the test instructions and did not see any issues. I did not do a code review.

getdave committed:
Improve function comment

Addresses https://github.com/WordPress/gutenberg/pull/36740#discussion_r755055265

getdave committed:
Amend ordering params to approximate current wp_nav_menus behaviour

noisysocks commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,92 @@ function block_core_navigation_render_submenu_icon() {
+	// see:
+	// - https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/nav-menu.php#L613-L619.
+	// - https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters.
+	$navigation_posts = get_posts(
```
By default `get_posts` won't return every single post. (I think the default is 5?) So the logic error in this function is that, if there are `n` navigation posts, and `n+1` is the non-empty one, this function will incorrectly return `null`.

We could fix this by setting `posts_per_page` to `-1` to return all posts, but this isn't a good idea as there could be more posts than the server has RAM.

We need to encode the _get the first non-empty post_ logic into the query itself. Unfortunately `WP_Query` doesn't have an argument which equates to `post_content != ''`. But that's ok, we know that in this case a non-empty navigation post will contain block markup so we could instead search for `<!--`.

Something like this:

```php
function block_core_navigation_get_non_empty_navigation() {
	$navigation_posts = get_posts(
			array(
 				'post_type'      => 'wp_navigation',
	 			'order'          => 'ASC',
	 			'orderby'        => 'name',
				'posts_per_page' => 1,
				's'              => '<!--',
	 		)
	);
	return count( $navigation_posts ) ? $navigation_posts[0] : null;
}
```

https://developer.wordpress.org/reference/classes/wp_query/parse_query/

getdave replied:
Great insight. I'll start work on this now. Thank you @noisysocks!

getdave replied:
This works well as far as I can see.

The original implementation was mirroring what is in `wp_nav_menu()`. The thought process was that it is highly unlikely that you will have 100s of `wp_navigation` posts so the in memory processing would probably never become an issue....probably.

Overall, so long as the `LIKE` in the query doesn't degrade the query performance we should be ok with your revised solution. It's certainly more elegant and avoids the looping that I had in the original implementation.

During my testing this seemed to work well. I appreciate your insight here 🙇‍♂️

spacedmonkey replied:
Let's not use get_posts where possible. `get_posts` is not cachable and does not run through filters.

noisysocks commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,92 @@ function block_core_navigation_render_submenu_icon() {
+ * @return array the array of blocks to be used as a fallback.
+ */
+function block_core_navigation_get_fallback() {
+
```
```suggestion
```

getdave committed:
Fetch Nav posts in query rather than filtering in memory

Implements suggestion from https://github.com/WordPress/gutenberg/pull/36740#discussion_r755647511

getdave committed:
Improve comments

getdave committed:
Spacing

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+			'post_type'      => 'wp_navigation',
+			'order'          => 'ASC',
+			'orderby'        => 'name',
+			'posts_per_page' => -1, // include all posts.
```
```suggestion
			'posts_per_page' => 1, // only the first post
```

adamziel replied:
As @noisysocks noticed, there may be more posts than RAM. This function returns only the first one anyway, so it doesn't have to process the other ones.

getdave replied:
Ah right I see. So we're querying first the first Nav Post which has block grammar inside it. Makes sense. At first I assumed it was a typo!

getdave committed:
Limit Nav Posts query to a single item

Co-authored-by: Adam Zielinski <adam@adamziel.com>

getdave committed:
Fix linting

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+ * @param array $parsed_blocks the parsed blocks to be normalized.
+ * @return array the normalized parsed blocks.
+ */
+function block_core_navigation_normalize_parsed_blocks( $parsed_blocks ) {
```
```suggestion
function block_core_navigation_filter_out_empty_blocks( $parsed_blocks ) {
```

"normalize" can mean many things, but "filter out empty blocks" tells me exactly what this function does.

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+ *
+ * @return WP_Post|null the first non-empty Navigation or null.
+ */
+function block_core_navigation_get_non_empty_navigation() {
```
```suggestion
function block_core_navigation_get_first_non_empty_navigation() {
```

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+		),
+	);
+
+	$navigation_post = block_core_navigation_get_non_empty_navigation();
```
```suggestion
	$navigation_post = block_core_navigation_get_first_non_empty_navigation();
```

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+
+	// Prefer using the first non-empty Navigation as fallback if available.
+	if ( $navigation_post ) {
+		$maybe_fallback = block_core_navigation_normalize_parsed_blocks( parse_blocks( $navigation_post->post_content ) );
```
```suggestion
		$maybe_fallback = block_core_navigation_filter_out_empty_blocks( parse_blocks( $navigation_post->post_content ) );
```

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -132,6 +132,83 @@ function block_core_navigation_render_submenu_icon() {
+ *
+ * @return array the array of blocks to be used as a fallback.
+ */
+function block_core_navigation_get_fallback() {
```
```suggestion
function block_core_navigation_get_fallback_blocks() {
```

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -204,29 +284,21 @@ function render_block_core_navigation( $attributes, $content, $block ) {
-			'blockName' => 'core/page-list',
-			'attrs'     => array(),
-		);
+		$fallback_blocks = block_core_navigation_get_fallback();
```
```suggestion
		$fallback_blocks = block_core_navigation_get_fallback_blocks();
```

adamziel commented on packages/block-library/src/navigation/index.php:
```diff
@@ -204,29 +284,21 @@ function render_block_core_navigation( $attributes, $content, $block ) {
-				return isset( $block['blockName'] );
-			}
-		);
+		$compacted_blocks = block_core_navigation_normalize_parsed_blocks( $parsed_blocks );
```
```suggestion
		$compacted_blocks = block_core_navigation_filter_out_empty_blocks( $parsed_blocks );
```

adamziel reviewed (approved):
I left a few suggestions related to naming, but otherwise this looks good to me! If you added a test or two, that would be absolutely stellar. Otherwise, 🚢

getdave committed:
Improve function naming

Co-authored-by: Adam Zielinski <adam@adamziel.com>

getdave commented:
> I left a few suggestions related to naming, but otherwise this looks good to me! If you added a test or two, that would be absolutely stellar. Otherwise, 🚢

Let's ship and I'll follow up tomorrow with e2e tests. I'll also sort out the Page List block attribute vs context thing.

bph commented:
@getdave I did s brief run through this before it was merged but didn't get to add my comments before it go merged. 
You had great test instructions! Thank you. 

After this sequence: 
- Open Site Editor.
- Remove any existing Nav block.
- Add Nav block from scratch. Don't add any items!
- Save your empty Nav block and publish the post.
- Go to front of site.
- See that the first of your Navigation Menus is rendered by default even though no Menu has been selected.

I did see the only menu I had was rendered. Putting myself in the shoes of a site owner or builder. I saw the menu, saw it had the wrong pages in there, so I clicked on the **Edit Site** link on top of my screen.  When the site editor loaded, I didn't know what to do next. My expectation would have been to see the menu now also in the editor, so I can make the changes, I wanted to make. 

I do like the default behavior. I just not sure what my expectation tells me about what happens next. Maybe this would work? 

Once the default menu is rendered on the frontend, because none was assigned before, that on subsequent Site editor views it could also be rendered there? 

Second observation: 
After I emptied all menus and trash them I didn't see a navigation rendered on the Front end. That worked. I looked into the Developer Tools and I saw a bunch of navigation html, though
![Screen Shot 2021-11-24 at 10 58 53 AM](https://user-images.githubusercontent.com/39980/143305195-15aa09cb-c579-43ef-9be9-6b118b5cfce2.png)

As if there is still an invisible navigation block. 
I didn't dive into this yet further, it could come from the theme template, though.

getdave commented:
Thanks @bph - appreciate you taking time to test this. I'll double check we can still fallback to Page List and the behaviour around what happens if Page List isn't available.

In terms of the UX we could show a warning on the block that it's currently showing a placeholder on the front end. I'm not confident we could land that in time for 5.9 though given everything else that's still outstanding.

We've also received some feedback / concerns about the fallback situation. I'd like to see how that plays out during the Beta period. It's simple enough that, if we find the majority of folks are not onboard with that, we can revert to simply rendering nothing.

Thanks again 🙇
